man/pj2if.1
.\" -*- NROFF -*-
.\" 
.TH "PJ2IF" "1" "" "Kevin Rosenberg" "Engineering"
.SH "NAME"
pj2if \- convert a projecttion file into an image (IF) file
.SH "SYNOPSIS"
.B pj2if in\-proj\-file out\-if\-file [OPTIONS]
.SH "DESCRIPTION "
\fIpj2if\fP converts \fIin\-proj\-file\fP, a file of projection data, into 
image (IF) file \fIout\-if\-file\fP, 
which can be further processed for viewing with \fIif2img\fP.
.SH "OPTIONS"
.TP 
.B \-\-verbose   
Verbose mode
.TP 
.B \-\-dump      
Dump all scan data
.TP 
.B \-\-version   
Print version
.TP 
.B \-\-help      
Print a terse help message
.SH "AUTHORS"
Kevin Rosenberg, M.D. <kevin@ctsim.org>
.SH "HISTORY"
CTSim was begun in 1983 using MS\-DOS and an EGA display adapter. In
1999 it was ported to GNU/Linux and later ported to Microsoft Windows.
.SH "SEE ALSO"
.BR ctsim (1)
.BR ctsimtext (1)
.BR if1 (1)
.BR if2 (1)
.BR ifinfo (1)
.BR linogram (1)
.BR phm2helix (1)
.BR phm2if (1)
.BR phm2pj (1)
.BR pjHinterp (1)
.BR pjinfo (1)
.BR pjrec (1)


NEWS
Version 4.2 Released Mar 2003

* Direct Fourier Reconstructions now supported

Version 4.0 Released Jan 2003

* Port to wxWindows 2.4

* Port to Mac OS X


Version 3.5 Released May 2002

* New reconstruction technique: Interpolate divergent beams to
  parallel beams for faster reconstructions.

* Multithreading for running length processes in background. Also
  takes advantage of multiple-CPU's in SMP systems for speed improvements
  with parallel processing.

* Region of interest can be specified for reconstruction zooming.

* Import PNG, PPM, and PGM images

* DICOM file import/export

* Clipboard for cutting and pasting images

* Can read and reconstruct SIEMENS AR.STAR DICOM raw data files

* Scattergram of T-Theta sampling. Useful for understanding divergent beam
  scanning.

* Startup-tips to help new users learn features.

* Ability to place detector at any arbitray position

* Histogram plotting of projection files


Version 3.0 New Features

* Greatly improved dialog boxes

* Online and print manuals with context-sensitive help!

* 3-d rotating views of image files

* Creation of arbitrary filter images

* Cubic interpolation for reconstructions

* All features of command line tools are now in graphical ctsim program!

* Complex-valued image files now supported, lots of image math
 functions added including Fourier transformations

* Visual and statistical image comparision functions

* Plotting of row and column data of single and comparison images

* Histogram Plotting of Image files.

* Conversion of projections to polar images

* Improved animation graphics

* New Microsoft Windows compatible self-installer with explorer extensions 
to CTSim files



Version 2.5 of CTSim has been released!

New Features of CTSim version 2.5:

- Now compiles with Microsoft Visual C++.

- First Microsoft Windows graphical user interface

- Bug fixes and modest documentation and program improvements over version 2.0



New Features of CTSim version 2.0:

- Entire code-base cleaned up, re-written, and converted from C to C++.

- Graphical user interface now included! Animation of projection and
  reconstruction processes. Displays projection and image files. Uses
  platform-independent wxWindows library so can run on UNIX and 
  Microsoft Windows platforms.

- Integrated G2 library for graphical display from command line tools
  (UNIX only)

- Added frequency-based filtering. Can provide significant speed-up
  in reconstructions compared to convolution-based filtering.

- Added highly-optimized backprojection method (idiff3)

- New command-line tools added for image statistics and comparisons with
  both text and graphical outputs.


tools/pj2if.cpp
/*****************************************************************************
** FILE IDENTIFICATION
**
**   Name:          pj2if.cpp
**   Purpose:       Convert an projection data file to an image file
**   Programmer:    Kevin Rosenberg
**   Date Started:  April 2000
**
**  This is part of the CTSim program
**  Copyright (C) 1983-2009 Kevin Rosenberg
**
**  This program is free software; you can redistribute it and/or modify
**  it under the terms of the GNU General Public License (version 2) as
**  published by the Free Software Foundation.
**
**  This program is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**  GNU General Public License for more details.
**
**  You should have received a copy of the GNU General Public License
**  along with this program; if not, write to the Free Software
**  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
******************************************************************************/

/* FILE
 *   pj2if.c                    Convert Raysum to image
 *
 * DATE
 *   Apr 1999
 */

#include "ct.h"
#include "timer.h"


enum { O_VERBOSE, O_HELP, O_VERSION, O_POLAR };

static struct option my_options[] =
{
  {"polar", 0, 0, O_POLAR},
  {"verbose", 0, 0, O_VERBOSE},
  {"help", 0, 0, O_HELP},
  {"version", 0, 0, O_VERSION},
  {0, 0, 0, 0}
};

static const char* g_szIdStr = "$Id$";

void
pj2if_usage (const char *program)
{
  std::cout << "usage: " << fileBasename(program) << " in-proj-file out-if-file [OPTIONS]\n";
  std::cout << "Converts a projection file to a imagefile\n";
  std::cout << std::endl;
  std::cout << "   --polar     Convert to polar format\n";
  std::cout << "   --verbose   Verbose mode\n";
  std::cout << "   --version   Print version\n";
  std::cout << "   --help      Print this help message\n";
}

int
pj2if_main (const int argc, char *const argv[])
{
  char *pj_name, *im_name;
  bool optVerbose = false;
  extern int optind;
  Timer timerProgram;

  while (1)
    {
      int c = getopt_long (argc, argv, "", my_options, NULL);
      if (c == -1)
        break;

      switch (c)
        {
        case O_VERBOSE:
          optVerbose = true;
          break;
        case O_VERSION:
#ifdef VERSION
          std::cout << "Version " << VERSION << std::endl << g_szIdStr << std::endl;
#else
          std::cout << "Unknown version number\n";
#endif
          return (0);
        case O_HELP:
        case '?':
          pj2if_usage(argv[0]);
          return (0);
        default:
          pj2if_usage(argv[0]);
          return (1);
        }
    }

  if (argc - optind != 2) {
    pj2if_usage(argv[0]);
    return (1);
  }

  pj_name = argv[optind];
  im_name = argv[optind + 1];

  Projections pj;
  if (! pj.read (pj_name)) {
    sys_error (ERR_SEVERE, "Can not open projection file %s", pj_name);
    return (1);
  }

  if (optVerbose) {
    std::ostringstream os;
    pj.printScanInfo (os);
    std::cout << os.str();
  }

  ImageFile im (pj.nDet(), pj.nView());

  ImageFileArray v = im.getArray();

  for (int iy = 0; iy < pj.nView(); iy++) {
    const DetectorArray& detarray = pj.getDetectorArray (pj.nView() - iy - 1);
    const DetectorValue* detval = detarray.detValues();
    for (int ix = 0; ix < pj.nDet(); ix++) {
      v[ix][iy] = detval[ix];
    }
  }

  im.labelAdd (pj.getLabel());
  im.labelAdd (Array2dFileLabel::L_HISTORY, "Conversion from .pj to .if", timerProgram.timerEnd());
  im.fileWrite (im_name);

  return(0);
}


#ifndef NO_MAIN
int
main (const int argc, char *const argv[])
{
  int retval = 1;

  try {
    retval = pj2if_main(argc, argv);
  } catch (exception e) {
    std::cerr << "Exception: " << e.what() << std::endl;
  } catch (...) {
    std::cerr << "Unknown exception\n";
  }

  return (retval);
}
#endif
